zkevm-circuits/src/blake2b/src/main.rs

```diff
@@ -0,0 +1,442 @@
+        region.assign_fixed(|| format!("Xor lookup table, row {}, the second", r), table[1], r, || Value::known(F::from(second)))?;
+        region.assign_fixed(|| format!("Xor lookup table, row {}, the result", r), table[2], r, || Value::known(F::from(first ^ second)))?;
+    }
+    return Ok(());
+}
+
+fn constrain_to_byte_xor<F:FieldExt>(meta: &mut ConstraintSystem<F>, table: &[Column<Fixed>; 3], 
+    selector: Column<Fixed>, first: ACell<F>, second: ACell<F>, result: ACell<F>) {
+        meta.lookup_any("Byte xor constraint", |meta| {
+            let selector = meta.query_fixed(selector, Rotation::cur());
+            return vec![
+                (selector.clone() * first.express(meta), meta.query_fixed(table[0], Rotation::cur())),
+                (selector.clone() * second.express(meta), meta.query_fixed(table[1], Rotation::cur())),
+                (selector.clone() * result.express(meta), meta.query_fixed(table[2], Rotation::cur()))];
+        });
+}
+
+fn set_result_and_activate<F:FieldExt, const N:usize >(region: &mut Region<F>, name: &str, 
+    row: usize, selector: Column<Fixed>, cells: &[ACell<F>; N], result: u128) -> Result<(), Error> {
+        set_byte_witnesses(region, &format!("{}, result", name), row, cells, result.to_le_bytes()[0..N].try_into().unwrap())?;     
+        region.assign_fixed(|| format!("{}, selector", name), selector, row, || Value::known(F::one()))?;
+        return Ok(());
+}
+
+#[derive(Debug, Copy, Clone)]
+struct ACell<F:FieldExt> {
+    pub column: Column<Advice>,
+    pub offset: u16,
+    _marker: PhantomData<F>
+}
+
+impl<F:FieldExt> ACell<F> {
+    fn express(&self, meta: &mut VirtualCells<F>) -> Expression<F> {
+        meta.query_advice(self.column, Rotation(self.offset as i32))
+    }
+}
+
+type U64ACells<F> = [ACell<F>; 8];
+
+#[derive(Debug, Clone)]
+struct AddGate<F:FieldExt, const S:usize>{
+    pub selector: Column<Fixed>,
+    pub sum: [ACell<F>; 9],
+}
+
+impl<F:FieldExt, const S:usize> AddGate<F, S> {
+    fn configure(meta: &mut ConstraintSystem<F>,
+                 selector: Column<Fixed>, 
+                 bytes: Option<Column<Fixed>>, 
+                 summands: &[&U64ACells<F>; S], 
+                 result: &U64ACells<F>,
+                 carry: ACell<F>) -> Self {
+        let mut sum = result.to_vec(); sum.push(carry);
+
+        meta.create_gate("AddGate", |meta| {
+            let selector = meta.query_fixed(selector, Rotation::cur());
+            let left = summands.iter().map(|s| byte_cells_to_number(meta, *s)).fold(gf(0), |e, s| e + s);
+            let right = byte_cells_to_number(meta, &sum);
+            return vec![selector * (left - right)];
+        });
+        
+        if let Some(b) = bytes { sum.iter().for_each(|c| constrain_to_set(meta, b, selector, *c)); }
+
+        return AddGate {selector, sum: sum.try_into().unwrap() };
+    }
+
+    fn put(&self, region: &mut Region<F>, row: usize, summands: &[u64; S]) -> Result<u64, Error>{
```

Comment:
```suggestion
    fn assign(&self, region: &mut Region<F>, row: usize, summands: &[u64; S]) -> Result<u64, Error>{
```

also just the more commonly used name everywhere else.